ledger-core/v2/virtual/execute/execute_test.go

```diff
@@ -197,3 +200,79 @@ func TestSMExecute_UpdateKnownRequests(t *testing.T) {
 		assert.Panics(t, checkerFunc, "panic with not implemented deduplication algorithm should be here")
 	}
 }
+
+func TestSMExecute_Semi_IncrementPendingCounters(t *testing.T) {
+	var (
+		mc  = minimock.NewController(t)
+		ctx = inslogger.TestContext(t)
+
+		prototype   = gen.UniqueReference()
+		caller      = gen.UniqueReference()
+		callee      = gen.UniqueReference()
+		outgoing    = gen.UniqueID()
+		objectRef   = reference.NewSelf(outgoing)
+		sharedState = &object.SharedState{
+			Info: object.Info{
+				KnownRequests:  make(map[reference.Global]struct{}),
+				ReadyToWork:    smsync.NewConditional(1, "ReadyToWork").SyncLink(),
+				MutableExecute: smsync.NewConditional(1, "MutableExecution").SyncLink(),
+			},
+		}
+	)
+
+	slotMachine := slotmachine.NewControlledSlotMachine(ctx, t, true)
+	slotMachine.PrepareMockedMessageSender(mc)
+	slotMachine.PrepareRunner(mc)
+
+	smExecute := SMExecute{
+		Payload: &payload.VCallRequest{
+			Polymorph:    uint32(payload.TypeVCallRequest),
+			CallType:     payload.CTConstructor,
+			CallFlags:    payload.BuildCallRequestFlags(contract.CallTolerable, contract.CallDirty),
+			CallOutgoing: outgoing,
+
+			Caller:              caller,
+			Callee:              callee,
+			CallSiteDeclaration: prototype,
+			CallSiteMethod:      "New",
+		},
+		Meta: &payload.Meta{
+			Sender: caller,
+		},
+	}
+
+	{
+		catalogWrapper := object.NewCatalogMockWrapper(mc)
+		var catalog object.Catalog = catalogWrapper.Mock()
+		slotMachine.AddInterfaceDependency(&catalog)
+
+		sharedStateData := smachine.NewUnboundSharedData(sharedState)
+		smObjectAccessor := object.SharedStateAccessor{SharedDataLink: sharedStateData}
+
+		catalogWrapper.AddObject(objectRef, smObjectAccessor)
+	}
+
+	{
+		pd := pulse.NewFirstPulsarData(10, longbits.Bits256{})
+		pulseSlot := conveyor.NewPresentPulseSlot(nil, pd.AsRange())
+		slotMachine.AddDependency(&pulseSlot)
+	}
+
+	slotMachine.Start()
+	defer slotMachine.Stop()
+
+	smWrapper := slotMachine.AddStateMachine(ctx, &smExecute)
+
+	require.Equal(t, uint8(0), sharedState.PotentialMutablePendingCount)
+	require.Equal(t, uint8(0), sharedState.PotentialImmutablePendingCount)
+
+	stepToWait := smWrapper.StateMachine().(*SMExecute).stepWaitObjectReady
+	if !slotMachine.StepUntil(smWrapper.WaitStep(stepToWait)) {
```

Comment:
Considering, that you have access to smExecute, I'd suggest to use it, instead of casting.
Also it may be beneficial to create explicit predicates, as you may want to wait for a step with flags, or for any step with specific migration, etc.

stepPredicate := slotMachine.IfStep(smExecute.stepWaitObjectReady)
or
stepPredicate := slotMachine.IfStepExt(SlotStep{...}).Not()
...

if !slotMachine.StepUntil(stepPredicate) {